add tutorial for PyTorch inference on AWS Graviton CPUs (#2719)
pytorch/tutorials: beginner_source/inference_on_aws_graviton_tutorial.rst

                  Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls
======================  ============  ============  ============  ============  ============  ============
           aten::addmm        95.61%        6.943s        97.10%        7.052s      23.507ms           300
       aten::clamp_min         2.31%     167.653ms         2.31%     167.653ms     838.265us           200
           aten::copy_         1.48%     107.593ms         1.48%     107.593ms     358.643us           300
     mymodel_inference         0.43%      31.167ms       100.00%        7.262s        7.262s             1
          aten::linear         0.04%       2.911ms        97.21%        7.060s      23.533ms           300
               aten::t         0.03%       2.414ms         0.07%       4.892ms      16.307us           300
            aten::relu         0.03%       2.281ms         2.34%     169.934ms     849.670us           200
======================  ============  ============  ============  ============  ============  ============

``Self CPU time total: 7.262s``

> This is around ``2x (7.262s vs 16.201s)`` performance improvement with the bfloat16 fastmath kernels. Next, let’s look at the smaller batch dimension scenario.

Reviewer: ```suggestion
This is around ``2x (7.262s vs 16.201s)`` performance improvement with the ``bfloat16`` fastmath kernels. Next, let’s look at the smaller batch dimension scenario.
```

Scenario2:  smaller batch dimension, for example, 32

.. code-block:: python

    X = torch.rand(32, 64, 64, device=device)
    with torch.set_grad_enabled(False):
        for _ in range(50):
            model(X) #Warmup
        with profile(activities=[ProfilerActivity.CPU]) as prof:
            with record_function("mymodel_inference"):
                for _ in range(100):
                    model(X)

    print(prof.key_averages().table(sort_by="self_cpu_time_total"))

> Following is the profiler output when the above script is run with the torch default configuration:

Reviewer: ```suggestion
You should see the following profiler output when the above script is run with the PyTorch default configuration:
```

                  Name    Self CPU %      Self CPU   CPU total %     CPU total   CPU time avg    # of Calls
======================  =============  ============  ============  ============  ============  ============
           aten::addmm        95.51%        5.821s        97.04%        5.914s      19.713ms           300
       aten::clamp_min         2.33%     142.244ms         2.33%     142.244ms     711.220us           200
           aten::copy_         1.51%      92.322ms         1.51%      92.322ms     307.740us           300
     mymodel_inference         0.45%      27.713ms       100.00%        6.094s        6.094s             1
          aten::linear         0.04%       2.495ms        97.16%        5.921s      19.736ms           300
               aten::t         0.03%       2.131ms         0.07%       4.441ms      14.803us           300
            aten::relu         0.03%       1.942ms         2.37%     144.186ms     720.930us           200
======================  =============  ============  ============  ============  ============  ============

``Self CPU time total: 6.094s``

> Following is the profiler output when run with the fast math mode enabled:

Reviewer: ```suggestion
The following output is the profiler output when run with the fast math mode enabled:
```

> ``$ export DNNL_DEFAULT_FPMATH_MODE=BF16``

Reviewer: ```suggestion
.. code-block:: bash

   $ export DNNL_DEFAULT_FPMATH_MODE=BF16``
```

======================  ============  ============  ============  ============  ============   =============
                Name     Self CPU %      Self CPU    CPU total %   CPU total    CPU time avg    # of Calls
======================  ============  ============  ============  ============  ============   =============
           aten::addmm        93.31%        3.848s        95.66%        3.944s      13.148ms           300
       aten::clamp_min         3.43%     141.309ms         3.43%     141.309ms     706.545us           200
           aten::copy_         2.33%      95.916ms         2.33%      95.916ms     319.720us           300
     mymodel_inference         0.67%      27.431ms       100.00%        4.123s        4.123s             1
          aten::linear         0.06%       2.471ms        95.83%        3.951s      13.170ms           300
               aten::t         0.05%       2.027ms         0.10%       4.243ms      14.143us           300
            aten::relu         0.05%       1.928ms         3.47%     143.237ms     716.185us           200
======================  ============  ============  ============  ============  ============   =============

> ``Self CPU time total: 4.123s``

Reviewer: ```suggestion
**Self CPU time total:** 4.123s
```

> This is around ``1.47x (4.123s vs 6.094s)`` improvement with fast math mode for the smaller batch dimensions. While the improvement is good, the performance is still not great. This is because of the runtime overhead (weights reorders and kernel launch time) from oneDNN and ACL backend outweighing the compute benefits from the ACL gemm kernels for the smaller batch compute.

Reviewer: ```suggestion
The fast math mode yields approximately a **1.47x  (4.123s vs 6.094s)**  for smaller batch dimensions. Although this improvement is noteworthy, the overall performance still leaves room for improvement. This is because of the runtime overhead (weights reorders and kernel launch time) from oneDNN and ACL backend outweighing the compute benefits from the ACL GEMM kernels for the smaller batch compute.
```

> 3. Improve inference performance with OpenBLAS for smaller batch dimensions

Reviewer: ```suggestion
Improve Inference Performance with OpenBLAS for Smaller Batch Dimensions
```

------------------------------------------------------------------------

> 

Reviewer: Can we explain first what is OpenBLAS and MKLDNN

> The inference performance for smaller batch dimensions can be improved by offloading the smaller shapes from MKLDNN to OpenBLAS backend. PyTorch wheels support OpenBLAS as the default BLAS library on AWS Graviton.Increase the threshold for mkldnn backend to offload the smaller batches to OpenBLAS. In the following example we use ``64`` as the threshold, so that input with ``batch dimension of 32`` is not dispatched to mkldnn, instead it is dispatched to OpenBLAS.

Reviewer: ```suggestion
The inference performance for smaller batch dimensions can be improved by offloading the smaller shapes from MKLDNN to OpenBLAS backend. PyTorch wheels support OpenBLAS as the default BLAS library on AWS Graviton. Increase the threshold for MKLDNN backend to offload the smaller batches to OpenBLAS. In the following example, we use ``64`` as the threshold, so that input with ``batch dimension of 32`` is not dispatched to MKLDNN. Instead, it is dispatched to OpenBLAS.
```

> ``$ export TORCH_MKLDNN_MATMUL_MIN_DIM=64``

Reviewer: ```suggestion
.. code-block:: bash

   $ export TORCH_MKLDNN_MATMUL_MIN_DIM=64
```

> Following is the output with OpenBLAS backend:

Reviewer: ```suggestion
Here is the output with the OpenBLAS backend:
```

======================  ============  ============  ============  =============  ============  =============
                  Name    Self CPU %      Self CPU   CPU total %     CPU total   CPU time avg    # of Calls
======================  ============  ============  ============  =============  ============  =============
           aten::addmm        96.25%        1.958s        97.51%        1.984s       6.612ms           300
       aten::clamp_min         1.28%      26.124ms         1.28%      26.124ms     130.620us           200
           aten::copy_         1.23%      24.951ms         1.23%      24.951ms      83.170us           300
     mymodel_inference         0.86%      17.423ms       100.00%        2.034s        2.034s             1
          aten::linear         0.08%       1.691ms        97.74%        1.988s       6.628ms           300
               aten::t         0.07%       1.520ms         0.14%       2.945ms       9.817us           300
            aten::relu         0.06%       1.258ms         1.35%      27.382ms     136.910us           200
======================  ============  ============  ============  =============  ============  =============

> ``Self CPU time total: 2.034s``

Reviewer: ```suggestion
**Self CPU time total:** 2.034s
```

> As seen here, switching to openBLAS improved the performance by ``2x (2.034s vs 4.123s)`` compared to the default configuration. This becomes significant for even smaller batch dimensions, for example, for a batch dimension of 10

Reviewer: ```suggestion
As you can see above, switching to openBLAS doubled the performance **(2.034s vs 4.123s)** compared to the default configuration. This becomes significant for even smaller batch dimensions, for example, for a batch dimension of 10:
```

.. code-block:: python

    X = torch.rand(10, 64, 64, device=device)
    with torch.set_grad_enabled(False):
        for _ in range(50):
            model(X) #Warmup
        with profile(activities=[ProfilerActivity.CPU]) as prof:
            with record_function("mymodel_inference"):
                for _ in range(100):
                    model(X)

    print(prof.key_averages().table(sort_by="self_cpu_time_total"))

> Following is the profiler output with oneDNN+ACL fast math mode:

Reviewer: ```suggestion
The following is the profiler output with oneDNN+ACL fast math mode:
```

                  Name    Self CPU %      Self CPU   CPU total %     CPU total   CPU time avg    # of Calls  
======================  ============  ============  ============  ============  =============  =============
           aten::addmm        87.81%        3.613s        91.90%        3.781s      12.604ms           300  
       aten::clamp_min         7.18%     295.437ms         7.18%     295.437ms       1.477ms           200  
           aten::copy_         4.07%     167.516ms         4.07%     167.516ms     558.387us           300  
     mymodel_inference         0.67%      27.708ms       100.00%        4.115s        4.115s             1  
          aten::linear         0.06%       2.499ms        92.06%        3.788s      12.627ms           300  
               aten::t         0.05%       1.982ms         0.11%       4.385ms      14.617us           300  
            aten::relu         0.05%       1.932ms         7.23%     297.369ms       1.487ms           200  
======================  ============  ============  ============  ============  =============  ============= 

> ``Self CPU time total: 4.115s``

Reviewer: ```suggestion
**Self CPU time total:** 4.115s
```

> and following is the profiler output with the openBLAS backend:

Reviewer: ```suggestion
and the following is the profiler output with the openBLAS backend:
```

> ``$ export TORCH_MKLDNN_MATMUL_MIN_DIM=64``

Reviewer: ```suggestion
.. code-block:: bash
   
   $ export TORCH_MKLDNN_MATMUL_MIN_DIM=64
```

======================  =============  ============  ============  ============  =============  ============   
                Name     Self CPU %      Self CPU     CPU total %   CPU total    CPU time avg    # of Calls
======================  =============  ============  ============  ============  =============  ============
           aten::addmm        92.66%        1.179s        95.23%        1.211s       4.038ms           300
       aten::clamp_min         2.83%      36.060ms         2.83%      36.060ms     180.300us           200
           aten::copy_         2.52%      32.013ms         2.52%      32.013ms     106.710us           300
     mymodel_inference         1.38%      17.521ms       100.00%        1.272s        1.272s             1
          aten::linear         0.14%       1.750ms        95.60%        1.216s       4.054ms           300
               aten::t         0.12%       1.475ms         0.24%       3.033ms      10.110us           300
            aten::relu         0.10%       1.285ms         2.94%      37.345ms     186.725us           200
======================  =============  ============  ============  ============  =============  ============

> ``Self CPU time total: 1.272s``

Reviewer: ```suggestion
**Self CPU time total:** 1.272s
```

> Here we observed ``3.2x (1.272s vs 4.115s)`` performance improvement by tuning the backend thresholds appropriately. We are working on making the backend selection automatic, with robust heuristics, for the future releases.

Reviewer: ```suggestion
Here we observed **3.2x (1.272s vs 4.115s)** performance improvement by tuning the backend thresholds appropriately. We are working on making the backend selection automatic, with robust heuristics, for the future releases.
```

> 4. Optimize memory allocation overhead with Linux Transparent huge pages

Reviewer: ```suggestion
Optimize Memory Allocation Overhead with Linux Transparent Huge Pages
```

-------------------------------------------------------------------------

> We also observed that for these larger networks, tensor memory allocations take significant portion of the inference latency. This can be optimized by enabling Linux transparent huge pages allocations from PyTorch C10 memory allocator. Set the following environment variable to enable it.

Reviewer: ```suggestion
We also observed that for these larger networks, tensor memory allocations take significant portion of the inference latency. This can be optimized by enabling Linux transparent huge pages allocations from PyTorch C10 memory allocator. Set the following environment variable to enable it:
```

> ``$ export THP_MEM_ALLOC_ENABLE=1``

Reviewer: ```suggestion
.. code-block:: bash

    $ export THP_MEM_ALLOC_ENABLE=1
```

> For the batch dimension of 256 and with fast math mode

Reviewer: ```suggestion
For the batch dimension of 256 and with fast math mode:
```

.. code-block:: python

    X = torch.rand(256, 64, 64, device=device)
    with torch.set_grad_enabled(False):
        for _ in range(50):
            model(X) #Warmup
        with profile(activities=[ProfilerActivity.CPU]) as prof:
            with record_function("mymodel_inference"):
                for _ in range(100):
                    model(X)

    print(prof.key_averages().table(sort_by="self_cpu_time_total"))

> Following is the profiler output with THP memory allocations enabled

Reviewer: ```suggestion
The following is the profiler output with THP memory allocations enabled:
```

.. table:: output with the fast math and thp memory allocations

======================  ============  ============  ============  ============  ==============  ============ 
                  Name   Self CPU %    Self CPU     CPU total %    CPU total     CPU time avg    # of Calls  
======================  ============  ============  ============  ============  ==============  ============
           aten::addmm        91.31%        6.115s        94.39%        6.321s      21.069ms           300  
       aten::clamp_min         4.82%     322.568ms         4.82%     322.568ms       1.613ms           200  
           aten::copy_         3.06%     204.602ms         3.06%     204.602ms     682.007us           300  
     mymodel_inference         0.61%      40.777ms       100.00%        6.697s        6.697s             1  
          aten::linear         0.05%       3.082ms        94.51%        6.329s      21.097ms           300  
            aten::relu         0.04%       2.547ms         4.85%     325.115ms       1.626ms           200  
======================  ============  ============  ============  ============  ==============  ============

> ``Self CPU time total: 6.697s``

Reviewer: ```suggestion
**Self CPU time total:** 6.697s
```

> This is an additional ``1.08x or 8% (6.697s vs 7.262s)`` improvement on top of the already optimized fast math mode measured above.

Reviewer: ```suggestion
This is an additional **1.08x or 8% (6.697s vs 7.262s)** improvement on top of the already optimized fast math mode measured above.
```

   .. code-block::

      python3 -m pip install torch

2. We will start by importing the required dependencies and defining the device will run on:

.. code-block:: python

    import torch
    import torch.nn as nn
    from torch.profiler import profile, record_function, ProfilerActivity

    # AWS Graviton3 cpu
    device = ("cpu")
    print(f"Using {device} device")

3. Given linear layers are at the heart of several neural networks, including transformers, we take a linear layer for this demo. We define our neural network by subclassing ``nn.Module``, and initializing the layers in ``__init__``. We construct the network with a typical large language model parameters to match the real world scenario:

.. code-block:: python

  class MyNeuralNetwork(nn.Module):
    def __init__(self):
        super().__init__()
        self.flatten = nn.Flatten()
        self.linear_relu_stack = nn.Sequential(
            nn.Linear(4096, 4096),
            nn.ReLU(),
            nn.Linear(4096, 11008),
            nn.ReLU(),
            nn.Linear(11008, 10),
        )

    def forward(self, x):
        x = self.flatten(x)
        logits = self.linear_relu_stack(x)
        return logits

> 4. Let's create an instance of MyNeuralNetwork, and move it to the device:

Reviewer: ```suggestion
4. Let's create an instance of ``MyNeuralNetwork``, and move it to the device:
```
